Fix: enable Multi-view as soon as second profile is started and show all profiles when it is active and further profiles are started

SlySven opened the pull request:
This will close #5268.

What was happening was that opening a third (or more) profile when multi-view was activated did not continue to show the second (or rather the currently active) one. This was due to an unconditional hiding of the profile that was active immediately before the new one was added. This needed to be gated by the flag that multi-view was active - so that if that was the case the previous profile was NOT then hidden.

Whilst working on this fix an intermediate solution resulted in segmentation faults when a profile was closed when multi-view was active. This seemed to be due to a `QResizeEvent` being received by a `TMainConsole` as its profile was being destroyed. A small change was put in to the event handler for this event so that it would be silently aborted should its `(Host*) TConsole::mpHost` pointer not be valid - this ought to reduce this sort of issue arising in the future.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

#### Release post highlight
Fix: multi-view mode controls are enabled as soon as a second profile is loaded, closing a small issue from June.

SlySven committed:
Fix: enable Multi-view as soon as second profile is started

This will close #5268.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented:
Unfortunately, this is not working right. Consider this:

1. Start Mudlet, random profile A
1. Open random profile B
1. Activate Multi View, this works now indeed
1. Open random profile C
1. Profile C will be shown in main window, but B is not shown anymore, only all three tabs are visible
1. Click profile B tab, now its pane will be shown in main window
1. Close profile B with black X button
1. Mudlet will crash!

edit: Bug 5 is in release 4.14.1 as well, but the crash is not

SlySven commented:
Okay, there are still things to be ironed out...

SlySven committed:
BugFix: prevent problems with more than two profiles

There were two issues:
1. Opening a third profile when multi-view was activated did not continue
to show the second one. This was due to an unconditional hiding of the
profile that was active immediately before the new one was added. This
needed to be gated by the flag that multi-view was active - so that if that
was the case the previous profile was NOT then hidden.

2. When more than two profiles were present and multi-view was active
closing one of them was causing a segmentation fault. This was because a
`TMainConsole::resizeEvent(...) was being activated in the profile that was
being closed and at that point the base classes `TConsole::mpHost` was no
longer safe to use. The fix seemed to be to run the check for that pointer
being non-null BEFORE allowing the `QResizeEvent` to be passed from the
derived `TMainConsole` class's overriding event handler to the base classes
one - whereas it was being done AFTER in the existing code.

With the first of these two tweaks it was found that there is no need to
forcibly clear the "Multiview" setting when Mudlet reverts to only having
one profile active. Instead it can be left active - but ineffective - until
another profile is loaded, at which point the multi-view mode takes effect
again.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
I think I have sorted them now - including "Bug 5"...

Kebap commented:
Thanks for the quick fix! 
Could you update title and first post here as well? 
I see you put a lengthy commit description already.

SlySven commented on src/mudlet.cpp:
```diff
@@ -1538,7 +1542,7 @@ void mudlet::addConsoleForNewHost(Host* pH)
 
     mpSplitter_profileContainer->addWidget(pConsole);
-    if (mpCurrentActiveHost) {
+    if (mpCurrentActiveHost && !mMultiView) {
```
This was what was causing the currently active profile to not be shown after a new one was opened when multi-view mode was enabled...!

Kebap reviewed (approved):
Does fix the bug now

SlySven commented:
> Could you update title and first post here as well?

Done so, but haven't mentioned the crashing as that was something induced by the first commit and thus no longer relevant.

SlySven committed:
Merge branch 'development' into Fix_enableMultiviewAsSoonAsSecondProfileStarted

vadi2 commented on src/mudlet.cpp:
```diff
@@ -1431,6 +1431,29 @@ void mudlet::closeHost(const QString& name)
+    }
+}
+
+void mudlet::reshowAllMainConsolesIfRequired()
```
```suggestion
void mudlet::reshowAllMainConsoles()
```

SlySven replied:
:-1: Again, the shorter wording implies something that is not the case.

demonnic replied:
I think most people will probably understand that it only does it if required, though? Unless there's another, very similar worded method name I don't know that we gain much by adding the "IfRequired" into the name?

SlySven replied:
The thing is, that the method makes an assessment as to whether it needs to "reshow" the main windows, but if it's assessment is wrong then it won't - the shorter name would suggest that it will always "reshow" them and I think that could be misleading.

keneanung replied:
The IfRequired is too imprecise for me. Maybe IfInMultiview? Again, I don't think this name is too long.

vadi2 replied:
We can make up ridiculously long names for all functions if we wanted to - there's plenty of conditions we can think of. `installPackage()` could be `installPackageIfPossible()`, `writeModule()` could be `startWritingModuleAsyncronously()` and so on. But clearly, the existing names have not been big problems that are bothering us - so we should keep using short and concise names. I'm not a fan of switching now.

SlySven replied:
Will `reshowRequiredMainConsoles()` please (or rather, at least, *satisfy*) everyone?

vadi2 replied:
Yeah

vadi2 commented on src/mudlet.cpp:
```diff
@@ -1431,6 +1431,29 @@ void mudlet::closeHost(const QString& name)
+    updateMultiviewControlsEnabled();
+}
+
+void mudlet::updateMultiviewControlsEnabled()
```
```suggestion
void mudlet::updateMultiviewControls()
```

SlySven replied:
:-1: To me the shorter wording implies that this method would modify the checked/unchecked state - which does need to happen on occasions - but is *not* done by this method.

vadi2 replied:
I don't think anyone will get massively confused by it and will make grievous errors. People will appreciate if they don't have to work with very long names, though.

demonnic replied:
How does the 'enabled' clear this up if it's left in place, though?

SlySven replied:
It is updating the "Enabled" states/flags of the multi-view controls - not (as might be assumed from the shorter name) the "Checked/Activated" ones....

keneanung replied:
I agree that not adding the enabled creates a wrong impression. However, the enabled confuses me here as well (which might be from my missing experience...). To me the name doesn't actually seem that long.

SlySven replied:
It was a *tad* shorter than `updateMultiviewControlsEnabledStates()`... 

or even `updateEnabledStatesOfMultiviewControls()` :man_shrugging:

SlySven replied:
:angry: So I listened to this discussion, made some changes - but you ignored me and changed this one afterwards to your own *overly short* (and I believe **misleading**) suggestion @vadi2 ...!

vadi2 replied:
I looked at `updateEnableOfMViewControls`, had no idea what it meant, thought of a shorter one and ended up arriving at the original without knowing it... sorry for overruling you, but we've got to keep stuff understandable for everyone involved. I don't believe I'm the only one who would say that `updateEnableOfMViewControls` is not a super intuitive name.

vadi2 reviewed (commented):
Looks good and I like the idea of more small functions over big and unwieldy ones. Just have a couple of suggestions for smaller names :)

vadi2 commented:
@SlySven ping

SlySven commented:
I really don't want to shorten those names - **do I *have* to?** :pleading_face:

Kebap commented:
Could we please have this bugfix in the next release and settle the internal naming conventions at another time? ⏲️

vadi2 commented:
No doubt it'll make it in before release, but it's important not to be shortsighted in this either.

SlySven commented:
Okay, I've gone for the following changes:
```diff
-    void updateMultiviewControlsEnabled();
+    void updateEnableOfMViewControls();

-    void reshowAllMainConsolesIfRequired();
+    void reshowRequiredMainConsoles();
```

SlySven committed:
Revise: shorten a couple of method names

I've shaved a few letters off of each.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented on src/mudlet.cpp:
```diff
@@ -1431,6 +1431,29 @@ void mudlet::closeHost(const QString& name)
+    updateEnableOfMViewControls();
+}
+
+void mudlet::updateEnableOfMViewControls()
```
```suggestion
void mudlet::updateMultiViewControls()
```

vadi2 commented on src/mudlet.cpp:
```diff
@@ -1503,28 +1526,9 @@ void mudlet::slot_tab_changed(int tabID)
-            dactionMultiView->setEnabled(false);
-        }
-    }
+    updateEnableOfMViewControls();
```
```suggestion
    updateMultiViewControls();
```

vadi2 commented on src/mudlet.cpp:
```diff
@@ -2838,6 +2842,7 @@ void mudlet::doAutoLogin(const QString& profile_name)
     emit signal_hostCreated(pHost, mHostManager.getHostCount());
     slot_connection_dlg_finished(profile_name, true);
     enableToolbarButtons();
+    updateEnableOfMViewControls();
```
```suggestion
    updateMultiViewControls();
```

vadi2 commented on src/mudlet.h:
```diff
@@ -331,6 +331,8 @@ class mudlet : public QMainWindow, public Ui::main_window
     void refreshTabBar();
     void updateDiscordNamedIcon();
+    // Has to be public as it needs to be called from dlgConnectionProfiles class:
+    void updateEnableOfMViewControls();
```
```suggestion
    void updateMultiViewControls();
```

vadi2 commented on src/dlgConnectionProfiles.cpp:
```diff
@@ -1689,6 +1689,8 @@ void dlgConnectionProfiles::loadProfile(bool alsoConnect)
         pHost->setupIreDriverBugfix();
     }
 
+    mudlet::self()->updateEnableOfMViewControls();
```
```suggestion
    mudlet::self()->updateMultiViewControls();
```

vadi2 commented on src/mudlet.cpp:
```diff
@@ -1431,6 +1431,29 @@ void mudlet::closeHost(const QString& name)
     int hostCount = mHostManager.getHostCount();
     emit signal_hostDestroyed(pH, --hostCount);
     mHostManager.deleteHost(pH->getName());
+    updateEnableOfMViewControls();
```
```suggestion
    updateMultiViewControls();
```

vadi2 reviewed (approved):
A further improvement... ! I hope it's ok.

vadi2 committed:
Apply suggestions from code review

vadi2 commented:
We didn't catch it, but this should have been an signal & a slot.
